### PR TITLE
Support functools.partial for event handlers

### LINKED ISSUE #766: Support functools.partial instances as event handlers
Currently you cannot pass an instance of functools.partial as an event handler since several places in helper.py assume the handler has a `__call__` and `__module__` property and can be passed to `inspect.getsource` (in `compute_fn_id` and `wrap_hander_with_event`). 

But you *can* pass a lambda, which is conceptually similar.

If this can't be supported for whatever reason, then maybe the event handler type hints should be changed since event handlers are annotated as `Callable` and a `partial` is a `Callable`.

### PR DESCRIPTION
**Pull request by wwwillchen**
Fixes #766

### CODE REVIEW

**Comment by malmaud** [comment]
I'm not a super-expert on the best way to do this but this seems reasonable to me.

**Comment by wwwillchen** [comment]
@richard-to WDYT about this feature / implementation? I think the feature is useful, but the implementation is a little bit hairy. In particular, we need to avoid getting a repr for types that return a memory address (e.g. vanilla Python classes)

```py
class Car:
  def __init__(self, name: str) -> None:
    self.name = name

car = Car(name="a") # repr -> <__main__.Car object at 0x79cfdc146080>
```

File: mesop/component_helpers/helper.py [hunk 1]
```diff
@@ -374,14 +380,63 @@ def register_event_handler(
   return fn_id
 
 
-@lru_cache(maxsize=None)
+def has_stable_repr(obj: Any) -> bool:
+  """Check if an object has a stable repr."""
```
Comment (richard-to) [inline mesop/component_helpers/helper.py:384]: [nit] Maybe add a bit more detail in this comment. Specifically would be good to know the potential limitations of using partial functions.

**Review by richard-to — APPROVED** [review]
Nice! I remember when I first started using Mesop, I had tried using functools.partial to no avail.

Looking at the code example, it does make the devex a lot nicer than having to pass in a key and then parsing the string.

Question: does this work when doing a dynamic loop? Say I wanted to add an event handler to each box using partial?

Yes, the limitation of needing a stable repr is interesting. How much of a drawback do you think that will be in practice?  I think it may be slightly annoying, but the improvement in devex is worth it I think (granted I haven't tested this out with some of the uses I have for this yet)

**Comment by wwwillchen** [comment]
> Nice! I remember when I first started using Mesop, I had tried using functools.partial to no avail.
>
> Looking at the code example, it does make the devex a lot nicer than having to pass in a key and then parsing the string.
> 
> Question: does this work when doing a dynamic loop? Say I wanted to add an event handler to each box using partial?

Yup. I added an example in the functools_partial example.

> Yes, the limitation of needing a stable repr is interesting. How much of a drawback do you think that will be in practice? I think it may be slightly annoying, but the improvement in devex is worth it I think (granted I haven't tested this out with some of the uses I have for this yet)

I think it's not too bad since this should cover most of the cases. It may be a little hard to understand for Mesop app developers, but I think we can work on improving the error message + docs if this becomes a common issue.

Thanks for the feedback, I'll merge this in.

**Pull request merged**